File: scripts/xhs/relationships.js
(params) => {
    const visible = el => !!el && el.getBoundingClientRect().width > 0 &&
        el.getBoundingClientRect().height > 0 && getComputedStyle(el).visibility !== 'hidden' &&
        Number(getComputedStyle(el).opacity || 1) > 0.1;
    if (visible(document.querySelector('.login-container'))) return {not_logged_in: true};
    const id = location.pathname.match(/^\/user\/profile\/([0-9a-fA-F]{24})\/?$/)?.[1];
    if (location.origin !== 'https://www.xiaohongshu.com' || id !== params.user_id) {
        return {error: '用户主页 ID 与目标不一致，已停止'};
    }
    const root = document.querySelector('.user-page > .user') || document.querySelector('.user-page .user-info');
    if (!root) return {ready: false};
    const names = [...root.querySelectorAll('.user-name')].filter(visible);
    if (names.length !== 1) return {ready: false};
    const nickname = names[0].textContent.trim();
    if (nickname !== params.expected_name) return {error: '用户主页昵称与预期不一致，已停止'};
    const norm = text => text.replace(/\s+/g, '');
    const toState = text => ['已关注', '互相关注', '相互关注'].includes(norm(text)) ? 'followed' :
        ['关注', '+关注', '回关'].includes(norm(text)) ? 'not-followed' : null;
    const buttons = [...root.querySelectorAll('button, [role="button"], .reds-button-new')]
        .filter(el => visible(el) && toState(el.textContent) !== null);
    const unique = buttons.filter(el => !buttons.some(other => other !== el && el.contains(other)));
    if (unique.length !== 1) return {ready: false};
    const button = unique[0];
    const state = toState(button.textContent);
    if (params.action === 'read') return {ready: true, user_id: id, nickname, state,
        label: button.textContent.trim()};
    if (params.action !== 'set') return {error: '未知关注操作'};
    if (state === params.desired_state) return {unchanged: true};
    if (!['followed', 'not-followed'].includes(params.desired_state) || state !== params.expected_state) {
        return {error: '点击前关注状态发生变化，已停止'};
    }
    if (button.disabled || button.getAttribute('aria-disabled') === 'true') {
        return {error: '关注按钮不可用，已停止'};
    }
    button.click();
    return {clicked: true};
}

File: scripts/xhs/relationships.py
"""关注状态查询和单次变更：核对用户、显式确认、刷新核验。"""

from __future__ import annotations

import json
import re
import time
from pathlib import Path
from typing import Any
from urllib.parse import urlencode

from .errors import NotLoggedInError, XHSError

_SCRIPT = Path(__file__).with_suffix(".js").read_text(encoding="utf-8")


class RelationshipError(XHSError):
    """关注关系读取或变更前校验失败。"""


def _run(page: Any, action: str, **params: Any) -> dict:
    value = page.evaluate(f"({_SCRIPT})({json.dumps({'action': action, **params})})")
    if not isinstance(value, dict):
        raise RelationshipError("无法读取用户主页的关注状态")
    if value.get("not_logged_in"):
        raise NotLoggedInError()
    if value.get("error") or value.get("__xhs_error"):
        raise RelationshipError(value.get("error") or value["__xhs_error"])
    return value


def _wait_profile(page: Any, user_id: str, expected_name: str, timeout: float = 12.0) -> dict:
    deadline = time.monotonic() + timeout
    while True:
        state = _run(page, "read", user_id=user_id, expected_name=expected_name)
        if state.get("ready"):
            return state
        if time.monotonic() >= deadline:
            raise RelationshipError("用户主页或关注按钮未就绪；可能无法访问或网页不支持该操作")
        time.sleep(0.3)


def set_follow(
    page: Any, user_id: str, expected_name: str, state: str,
    *, xsec_token: str = "", confirmed: bool = False, timeout: float = 12.0,
) -> dict:
    """默认只读预览；确认后至多点击一次，刷新确认最终状态，不自动重试。"""
    if not re.fullmatch(r"[0-9a-fA-F]{24}", user_id):
        raise RelationshipError("用户 ID 必须为 24 位十六进制 ID")
    if not expected_name.strip():
        raise RelationshipError("必须提供目标用户完整昵称")
    if state not in ("followed", "not-followed"):
        raise RelationshipError("目标状态必须为 followed 或 not-followed")
    url = f"https://www.xiaohongshu.com/user/profile/{user_id}"
    if xsec_token:
        url += "?" + urlencode({"xsec_token": xsec_token, "xsec_source": "pc_note"})
    page.navigate(url)
    page.wait_for_load()
    current = _wait_profile(page, user_id, expected_name)
    base = {"user_id": user_id, "nickname": expected_name, "desired_state": state}
    if not confirmed:
        return {**base, "success": True, "status": "preview", "changed": False,
                "current_state": current["state"], "would_change": current["state"] != state}
    if current["state"] == state:
        return {**base, "success": True, "status": "unchanged", "changed": False,
                "current_state": current["state"]}
    # 身份、当前状态、目标状态检查和点击在一次脚本中完成；点击后故障不能视为可重试。
    try:
        clicked = _run(page, "set", user_id=user_id, expected_name=expected_name,
                       expected_state=current["state"], desired_state=state)
        if clicked.get("unchanged"):
            return {**base, "success": True, "status": "unchanged", "changed": False,
                    "current_state": state}
        if not clicked.get("clicked"):
            raise RelationshipError("网页未确认已经执行关注操作")
        deadline = time.monotonic() + timeout
        while True:
            after = _run(page, "read", user_id=user_id, expected_name=expected_name)
            if after.get("ready") and after.get("state") == state:
                # 防止把页面乐观更新误当成持久化成功。
                page.navigate(url)
                page.wait_for_load()
                reloaded = _wait_profile(page, user_id, expected_name)
                if reloaded.get("state") == state:
                    return {**base, "success": True, "status": "updated", "changed": True,
                            "current_state": state, "verification": "reloaded_page"}
                break
            if time.monotonic() >= deadline:
                break
            time.sleep(0.3)
    except Exception:
        return {**base, "success": False, "status": "unknown", "changed": None,
                "error": "操作后的状态无法确认；请核对主页，勿直接重试"}
    return {**base, "success": False, "status": "unknown", "changed": None,
            "error": "未能确认刷新后的目标状态；请核对主页，勿直接重试"}
